build,ImageBuilder: use kmod archive

aparcar opened the pull request:
Use kmod archive created by the Buildbots instead of storing them locally, this saves about 10MB per ImageBuilder.

The referenced step in `buildbot.git` should be removed.

aparcar committed:
build: FeedSourcesAppend add kmod feed to snapshot

Nightly snapshot builds of OpenWrt change their kernels versions
frequently and lose thereby compatibility to kmods from the upstream
target specific packages feed.

To allow opkg to install packages over multiple days a kmod archive is
offered at $target/$subtarget/kmods/$kernelversion and added as a feed
to created snapshot images via a buildbot step[1].

Instead of using a buildstep add the kmod feed directly via
FeedSourcesAppend to be included in the ImageBuilder repositories.conf
as well. This is conditionally only done for SNAPSHOT builds and when
running as BUILDBOT. Releases are unaffected as they don't include
kernel version changes and local builds may use different kernel
versions or magics than available upstream.

This commit allows in a future step to ship ImageBuilders without a
locally stored kmod archive.

[1]: https://git.openwrt.org/?p=buildbot.git;a=blob;f=phase1/master.cfg;h=3ba7a1606e89b095b10555e703ea96e93295deec;hb=HEAD#l1025

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar committed:
build,IB: include kmods only in local builds

The buildbots generate a kmod archive which should be used instead of a
local copy. This is possible due to the introduction of a kernelversion
specific feed.

This commit adds the ability of using only signed package feeds.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar commented:
@jow- please have a look, I bet you know a better way to implement this.

jow- commented:
Please runtime test using the openwrt-19.07 branch HEAD (not a release tag) as well

aparcar commented:
```
## Place your custom repositories here, they must match the architecture and version.
# src/gz %n http://downloads.openwrt.org/releases/19.07-SNAPSHOT
# src custom file:///usr/src/openwrt/bin/ath79/packages

## Remote package repositories
src/gz openwrt_core http://downloads.openwrt.org/releases/19.07-SNAPSHOT/targets/ath79/generic/packages
src/gz openwrt_base http://downloads.openwrt.org/releases/19.07-SNAPSHOT/packages/mips_24kc/base
src/gz openwrt_kmods http://downloads.openwrt.org/releases/19.07-SNAPSHOT/targets/ath79/generic/kmods/4.14.195-1-unknown

## This is the local package repository, do not remove!
src imagebuilder file:packages
```

Seem to work. I didn't compile the whole things so the kernel version is unknown.

aparcar commented:
Related https://bugs.openwrt.org/index.php?do=details&task_id=3363&order=dateopened&sort=desc

aparcar committed:
opkg: fix kmod selection

This patch fixes the kmod selection if multiple versions are available.

Signed-off-by: Paul Spooren <mail@aparcar.org>

aparcar commented:
OPKG patch added

aparcar commented:
Without the `opkg` patch:

```
root@0020fb6c63d0:/# opkg update
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_core
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/packages/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/kmods/5.4.63-1-2b15fbe2fc4637fac17bbc326e475cff//Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_kmods
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/kmods/5.4.63-1-2b15fbe2fc4637fac17bbc326e475cff//Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/snapshots/packages/x86_64/base/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_base
Downloading https://downloads.openwrt.org/snapshots/packages/x86_64/base/Packages.sig
Signature check passed.
root@0020fb6c63d0:/# opkg install kmod-tun
Installing kmod-tun (5.4.71-1) to root...
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/packages/kmod-tun_5.4.71-1_x86_64.ipk
Collected errors:
 * satisfy_dependencies_for: Cannot satisfy the following dependencies for kmod-tun:
 * 	kernel (= 5.4.71-1-f01ee92d3ce1b58ffedaf81f93d036df)
 * opkg_install_cmd: Cannot install package kmod-tun.
root@0020fb6c63d0:/#
```
With `opkg` patch:

```
root@9134b45523a4:/tmp# opkg update
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/packages/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_core
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/packages/Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/kmods/5.4.63-1-2b15fbe2fc4637fac17bbc326e475cff//Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_kmods
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/kmods/5.4.63-1-2b15fbe2fc4637fac17bbc326e475cff//Packages.sig
Signature check passed.
Downloading https://downloads.openwrt.org/snapshots/packages/x86_64/base/Packages.gz
Updated list of available packages in /var/opkg-lists/openwrt_base
Downloading https://downloads.openwrt.org/snapshots/packages/x86_64/base/Packages.sig
Signature check passed.
root@9134b45523a4:/tmp# opkg install kmod-tun
Installing kmod-tun (5.4.63-1) to root...
Downloading https://downloads.openwrt.org/snapshots/targets/x86/64/kmods/5.4.63-1-2b15fbe2fc4637fac17bbc326e475cff//kmod-tun_5.4.63-1_x86_64.ipk
Configuring kmod-tun.
root@9134b45523a4:/tmp# 
```

aparcar commented:
Having this merged allows remote kmod archives in the ImageBuilder, which also solves the "signed" ImageBuilder issue because the unsigned local `packages/` feed is no longer required.

aparcar committed:
build,IB: reload packages/ only if existing

With the fix of external kmod feeds it is possible to ship the
ImageBuilder without any packages except the pseudo packages kernel and
libc. Therefore the local package feeds becomes optional.

This commit adds a check to the package_reload function to only run if
the local feed is existing.

Signed-off-by: Paul Spooren <mail@aparcar.org>

dangowrt commented:
pulled to my staging tree for testing. thanks!

aparcar commented:
Updated the opkg patch so that it only resolves packages if they are the right architecture and not already added.
